### PR TITLE
[perf] avoid HunyuanImage3 timestep scalar sync

### PR DESCRIPTION
**Pull request by TaffyOfficial**
## Purpose

Remove an avoidable GPU synchronization in the HunyuanImage3 AR embedding path.

`embed_input_ids()` used `timestep_mask.sum().item()` to count `<timestep>` tokens before replacing those positions with `timestep_emb(0)`. That count is not needed: every `<timestep>` position receives the same embedding value.

This PR computes `timestep_emb(0)` once and broadcasts it with `torch.where(...)`, so the replacement stays on GPU and keeps the same output semantics.

## Benchmark Details

The performance numbers below were collected on the AR-only graph serving path:

- Server path: `vllm serve` with the HunyuanImage3 AR deploy config and `enforce_eager: false`.
- Benchmark script: `vllm bench serve`.
- Endpoint/backend: `/v1/chat/completions` with `openai-chat-omni`.
- Workload: `random-mm`, 10 prompts, max concurrency 1, one 1024x1024 image per request, input length 256, output length 512, `temperature=0`, `ignore_eos=True`.
- Request body override: `{"modalities":["text"],"bot_task":"think"}`.

Benchmark command shape:

```bash
vllm bench serve \
  --omni \
  --backend openai-chat-omni \
  --host 127.0.0.1 \
  --port <serve-port> \
  --endpoint /v1/chat/completions \
  --model tencent/HunyuanImage-3.0-Instruct \
  --dataset-name random-mm \
  --num-prompts 10 \
  --request-rate inf \
  --max-concurrency 1 \
  --random-input-len 256 \
  --random-output-len 512 \
  --random-mm-base-items-per-request 1 \
  --random-mm-limit-mm-per-prompt '{"image":1,"video":0,"audio":0}' \
  --random-mm-bucket-config '{"(1024, 1024, 1)": 1.0}' \
  --extra-body '{"modalities":["text"],"bot_task":"think"}' \
  --temperature 0 \
  --ignore-eos
```

The before/after profiling wrapper only added timing probes around the same serving benchmark. It did not change the request payload.

Profiling showed the targeted sync point was removed, and the same benchmark improved end to end:

| Metric | Before | After | Change |
| --- | ---: | ---: | ---: |
| Total `Tensor.item()` time | 50.25 s | 0.36 s | -99.3% |
| Slow `Tensor.item() >50ms` | 18 | 0 | -100.0% |
| Max `Tensor.item()` | 209.153 ms | 10.174 ms | -95.1% |
| Input-prep p50 | 5.910 ms | 1.642 ms | -72.2% |
| Benchmark duration | 41.83 s | 40.20 s | -3.9% |
| Output tok/s | 72.54 | 75.47 | +4.0% |

The `Tensor.item()` sync cost drops by 99.3%, and the same benchmark improves by about 4.0% end to end. This PR only addresses the `<timestep>` scalar sync; other input-prep costs, such as image encoding and prompt/prefill work, are unchanged.

## Accuracy Evaluation

This change should not affect model accuracy: every `<timestep>` token still receives the same `timestep_emb(0)` value as before, and all non-`<timestep>` token embeddings are preserved.

I also ran an AR-only accuracy check instead of only relying on the unit test. This check reuses the same CoT text metrics from `tests/e2e/accuracy/test_hunyuan_image3.py`, but stops at the AR text output because this PR only changes AR token embedding replacement and does not modify the DiT image generation path.

- Prompt/images: same `PROMPT` and `TEST_IMAGE_URLS` as the HunyuanImage3 accuracy test.
- Prompt format: `task=it2i`, `bot_task=think_recaption`, `sys_type=en_unified`, `num_images=2`.
- Metrics:
  - `cot_semantic_sim`: BGE-M3 dense semantic similarity against `COT_REF`, threshold `>=0.9`.
  - `text_prefix_match_count`: number of leading characters that exactly match `COT_REF`, threshold `>=10`.

These metrics are useful here because a wrong `<timestep>` embedding replacement would change the AR hidden states before generation and can show up either as an early CoT prefix mismatch or as a lower full-CoT semantic similarity score. The passing result means the patched AR path still produces semantically matching reasoning text on the same prompt/images used by the accuracy test.

Result:

| Metric | Value | Threshold | Result |
| --- | ---: | ---: | --- |
| `cot_semantic_sim` | 0.9718 | 0.9 | PASS |
| `text_prefix_match_count` | 29 | 10 | PASS |
| AR output length | 1181 chars | N/A | recorded |

This is the relevant accuracy check for this PR because the patch only changes the AR embedding path and does not touch the DiT image generation stage.

## Test Plan

Unit regression:

```bash
python -m pytest tests/diffusion/models/hunyuan_image3/test_hunyuan_image3_sampler.py -q
```

Static checks:

```bash
ruff check \
  vllm_omni/model_executor/models/hunyuan_image3/hunyuan_image3.py \
  tests/diffusion/models/hunyuan_image3/test_hunyuan_image3_sampler.py

ruff format --check --diff \
  vllm_omni/model_executor/models/hunyuan_image3/hunyuan_image3.py \
  tests/diffusion/models/hunyuan_image3/test_hunyuan_image3_sampler.py

python -m py_compile \
  vllm_omni/model_executor/models/hunyuan_image3/hunyuan_image3.py \
  tests/diffusion/models/hunyuan_image3/test_hunyuan_image3_sampler.py

git diff --check
```

Performance benchmark:

```bash
vllm serve tencent/HunyuanImage-3.0-Instruct \
  --omni \
  --port <serve-port> \
  --trust-remote-code \
  --deploy-config <hunyuan_image3_ar_graph.yaml> \
  --chat-template <hunyuan_image3_i2t.jinja> \
  --chat-template-content-format openai
```

Then run the `vllm bench serve` command shown in **Benchmark Details**.

AR-only accuracy check:

```bash
python /tmp/hy3_ar_prefix_accuracy.py
```

The helper loads the same `PROMPT`, `TEST_IMAGE_URLS`, `COT_REF`, `cot_semantic_sim`, and `text_prefix_match_count` metric definitions as `tests/e2e/accuracy/test_hunyuan_image3.py`, then runs only the AR text path because this PR does not touch DiT image generation.

The corresponding full-pipeline accuracy tests that own these metrics are:

```bash
python -m pytest tests/e2e/accuracy/test_hunyuan_image3.py::test_image_to_image_alignment -q
python -m pytest tests/e2e/accuracy/test_hunyuan_image3.py::test_image_to_image_alignment_online -q
```

## Test Result

- PASS: HunyuanImage3 timestep embedding regression: `18 passed`.
- PASS: lint/static checks.
- PASS: AR-only HunyuanImage3 CoT semantic accuracy: `cot_semantic_sim=0.9718`, threshold `>=0.9`.
- PASS: AR-only HunyuanImage3 prefix accuracy: `text_prefix_match_count=29`, threshold `>=10`.

### CODE REVIEW

**Comment by TaffyOfficial** [comment]
@Bounty-hunter

**Comment by Bounty-hunter** [comment]
Nice catch，can you please clarify in the PR description which benchmark scripts were used and paste the corresponding results? 
It is also recommended to run both local test performance and accuracy  evaluations. In theory, this change should not have any impact on model accuracy.

**Comment by TaffyOfficial** [comment]
> Nice catch，can you please clarify in the PR description which benchmark scripts were used and paste the corresponding results? It is also recommended to run both local test performance and accuracy evaluations. In theory, this change should not have any impact on model accuracy.

fix

**Comment by TaffyOfficial** [comment]
﻿Follow-up context on where this came from and what is still left:

This was found while looking at AR graph online profiling traces for decode-step GPU idle time. The original visual symptom looked like short memcpy events followed by an idle gap, but the copy itself was only a small part of the window. After drilling into the CPU stack for the idle window, the repeated slow records pointed to `embed_input_ids()` calling `timestep_mask.sum().item()`.

That `.item()` reads a GPU scalar on the CPU side every decode step, which forces a synchronization before the CUDA graph replay/input-prep path can continue. Since the count is only used to decide whether to replace `<timestep>` positions, the replacement can stay on GPU by computing `timestep_emb(0)` once and applying `torch.where(...)`.

What this PR covers:

- removes the confirmed per-step scalar sync in the HunyuanImage3 AR embedding path;
- keeps the same semantics for `<timestep>` and non-`<timestep>` token embeddings;
- shows the targeted profiler cost drop and a small no-profiler benchmark improvement;
- validates AR text accuracy with the same CoT semantic/prefix metrics used by the HunyuanImage3 accuracy test.

What is still not claimed as fixed:

- remaining tail gaps around CUDA graph launch/replay still exist in some traces;
- input-prep / multimodal encoder wait may still have separate costs;
- async output, sampling, and TP-rank control-plane skew are still possible follow-up areas.

So the scope here is intentionally narrow: this PR fixes the confirmed `<timestep>` scalar synchronization. The remaining idle regions need separate profiling and should not be attributed to this `.item()` after this patch.

**Comment by hsliuustc0106** [comment]
Reviewed PR #4363: perf: avoid HunyuanImage3 timestep scalar sync

**Review by Bounty-hunter — APPROVED** [review]
LGTM

**Comment by TaffyOfficial** [comment]
@Gaohan123 @hsliuustc0106

**Review by hsliuustc0106 — APPROVED** [review]
lgtm

**Comment by NilouBao** [comment]
Hi, I used the same command as you to test the HunyuanImage-3.0 model deployed with vLLM-Omni (I tried both AR-only and AR+DiT modes), but more than half of the requests failed, and the server did not receive any requests. Have you encountered a similar issue?
